### PR TITLE
FastSmrMapLoader checkpoint/trim support

### PR DESCRIPTION
**Pull request by rogermichoud**
The FastSmrMapLoader is now able to load trimm/checkpoint. Here are the 3 steps of
the recovery:
1. Search for the checkpoints (for now, we have to read the whole log to find them)
2. Apply the checkpoints update in parallel
3. Go through the the whole log and apply only the entries after the checkpoints

### CODE REVIEW

**Comment by rogermichoud** [comment]
I have several dependency before it can be merged:

1. TrimStart API
2. Bulk read support trim exception
3. Bulk read support non-contiguous addresses (not only range)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **c57d11b**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #758 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/c4ZP2gIXrJL2wQJQEpEpL5HlKU6J6U0J)

File: runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java [hunk 1]
```diff
@@ -288,49 +453,95 @@ public void loadMaps() {
...
+
+    /**
+     * This method will apply for each address the consumer given in parameter.
+     *
+     * The Necromancer thread is used to do the heavy lifting.
+     * @param logDataProcessor
+     */
+    private void applyForEachAddress(BiConsumer<Long, ILogData> logDataProcessor) {
+
+        summonNecromancer();
```
Comment (dahliamalkhi) [inline runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java]: I think you need to change nextRead = 0 to nextRead = logHead

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **bfd541f**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #758 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/NZdSIEEE3RZB8O8rnwtjd06e6AP7rE0T)

File: runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java [hunk 2]
```diff
@@ -70,11 +71,14 @@
...
     private int timeoutInMinutesForLoading = DEFAULT_TIMEOUT_MINUTES_FAST_LOADING;
 
-
     @Setter
     @Getter
     private boolean recoverSequencerMode;
 
+    @Setter
+    @Getter
+    private boolean iAmSureMyLogHasNoCheckPoint = false;
```
Comment (rogermichoud) [inline runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java]: Changed it.

File: runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java (hunk 1 above)
Comment (rogermichoud) [inline runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java]: Yes, I missed that one! Fixed

File: format/proto/types.proto [hunk 3]
```diff
@@ -31,6 +31,8 @@ message LogEntry {
     optional CheckpointEntryType checkpointEntryType = 10;
     optional int64 checkpointID_most_significant = 11;
```
Comment (no2chem) [inline format/proto/types.proto]: ID -> Id

Comment (rogermichoud) [inline format/proto/types.proto]: should I do this change in all my "ID" ?

**Comment by no2chem** [comment]
Yeah, the new style rules don’t allow ID as a word

On Thu, Jun 22, 2017 at 8:18 PM, Roger <notifications@github.com> wrote:
@rogermichoud commented on this pull request.




In format/proto/types.proto [https://github.com/CorfuDB/CorfuDB/pull/758#discussion_r123664874] :

> @@ -31,6 +31,8 @@ message LogEntry {
     optional CheckpointEntryType checkpointEntryType = 10;
     optional int64 checkpointID_most_significant = 11;


should I do this change in all my "ID" ?

**Comment by rogermichoud** [comment]
okay, let me address this.

**Comment by rogermichoud** [comment]
I changed most of the one I found. I didn't take the call with the risky one (like CorfuRuntime.getStreamID). I think this should be addressed in another commit. If we change this API, it will probably break MP.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **1494851**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #758 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/RjR118Up6LtFnygIUHABiTFYzh115hOT)

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **23c9018**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #758 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/SgnvWL7SLOtoSkgGSyTkKWZKdCLJF8HX)

**Comment by rogermichoud** [comment]
@no2chem I had to rebase and fix conflict because I changed all the ID -> id in many files. Could you merge it before there is another merge conflict ?

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **f409d17**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #758 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/RYMy03EJziRzKzxBfYaQ5lJbH5dRYlhW)

**Pull request merged**